src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp

```diff
@@ -0,0 +1,300 @@
+    {
+        return implementation_buffer;
+    }
+    else
+    {
+        if (!file_size)
+            file_size = implementation_buffer->getFileSize();
+
+        auto implementation_buffer_creator = [object, read_settings, read_hint, file_size, this]()
+        {
+            return std::make_unique<BoundedReadBuffer>(object_storage->readObject(object, read_settings, read_hint, file_size));
+        };
+
+        FileCache::Key key = getCacheKey(object.getPathKeyForCache());
+        LOG_TEST(log, "Reading from file `{}` with cache key `{}`", object.absolute_path, key.toString());
+        return std::make_unique<CachedOnDiskReadBufferFromFile>(
+            object.absolute_path,
+            key,
+            cache,
+            implementation_buffer_creator,
+            read_settings,
+            CurrentThread::isInitialized() && CurrentThread::get().getQueryContext() ? std::string(CurrentThread::getQueryId()) : "",
+            file_size.value(),
+            /* allow_seeks */true,
+            /* use_external_buffer */false);
+    }
+}
+
+
+std::unique_ptr<WriteBufferFromFileBase> CachedObjectStorage::writeObject( /// NOLINT
+    const StoredObject & object,
+    WriteMode mode, // Cached doesn't support append, only rewrite
+    std::optional<ObjectAttributes> attributes,
+    FinalizeCallback && finalize_callback,
+    size_t buf_size,
+    const WriteSettings & write_settings)
+{
+    auto modified_write_settings = IObjectStorage::patchSettings(write_settings);
+    auto implementation_buffer = object_storage->writeObject(object, mode, attributes, std::move(finalize_callback), buf_size, modified_write_settings);
+
+    bool cache_on_write = fs::path(object.absolute_path).extension() != ".tmp"
+        && modified_write_settings.enable_filesystem_cache_on_write_operations
+        && FileCacheFactory::instance().getSettings(cache->getBasePath()).cache_on_write_operations;
+
+    auto path_key_for_cache = object.getPathKeyForCache();
+    /// Need to remove even if cache_on_write == false.
+    removeCacheIfExists(path_key_for_cache);
+
+    if (cache_on_write)
+    {
+        auto key = getCacheKey(path_key_for_cache);
+        LOG_TEST(log, "Caching file `{}` to `{}` with key {}", object.absolute_path, getCachePath(path_key_for_cache), key.toString());
+
+        return std::make_unique<CachedOnDiskWriteBufferFromFile>(
+            std::move(implementation_buffer),
+            cache,
+            implementation_buffer->getFileName(),
+            key,
+            modified_write_settings.is_file_cache_persistent,
+            CurrentThread::isInitialized() && CurrentThread::get().getQueryContext() ? std::string(CurrentThread::getQueryId()) : "",
+            modified_write_settings);
+    }
+
+    return implementation_buffer;
+}
+
+void CachedObjectStorage::removeCacheIfExists(const std::string & path_key_for_cache)
+{
+    if (path_key_for_cache.empty())
```

Comment:
Easy, this parameter is optional. For example, for LocalObjectStorage if object does not exist, it is not possible to get the cache key because inode number is used to construct it.